### PR TITLE
Add graphql AST helpers | Generate cell input props

### LINKED ISSUE #2811: Expose types from Cell when they're used
We're not currently generating types for when we _use_ Cells. i.e. the Cell input props, which are used to make the graphql query.

### PR DESCRIPTION
**Pull request by jtoar**
The `@redwoodjs/internal` package has some Redwood-specific AST helpers; @peterp thought it'd be a good idea to add some GraphQL ones to the mix. I kind of just got my feet wet here, trying to mimic the [jsx one](https://github.com/redwoodjs/redwood/blob/main/packages/internal/src/jsx.ts) that Peter recently added. 

All the helper does (that this PR introduces) is parse an already-parsed GraphQL query (i.e. uses the gql template literal) into something more friendly. I think I'll have to add a lot more properties, or just change my approach entirely (smaller functions?). It's probably best to work backwards from a specific use case. 

But I'd appreciate feedback in the meantime! Especially regarding my TypeScript, there's one type I haven't worked out yet.

Update: adding some commits to use this functionality for Cell mirror generation
Fixes #2811

### CODE REVIEW

**Comment by peterp** [comment]
This is looking good! We need to modify the ast finder to search for template literals that are surrounded by `gql`, I can help with that!

**Comment by dac09** [comment]
Hey @jtoar now have a need for this. Would it be possible to have a helper function just to extract the operation name? 

Want to use it to parse gql in Cells, so I know which types to import in mirror cells.

File: packages/internal/src/gql.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+  return operations
+}
+
+const getFields = (field: FieldNode): any => {
+  // base
+  if (!field.selectionSet) {
+    return field.name.value
+  } else {
+    const obj = {
```
Comment (dac09) [inline packages/internal/src/gql.ts:46]:
```suggestion
    const obj: Record<string, FieldNode[]>
```
This fixes the type error :)

**Comment by jtoar** [comment]
> Would it be possible to have a helper function just to extract the operation name?

For sure! Should be pretty easy I think

**Comment by dac09** [comment]
Update: adding some commits to use this functionality for Cell mirror generation. I thought it made more sense to have it in this PR, than cause merge conflict havoc by having a separate one that relies on this PR

**Comment by dac09** [comment]
> This is looking good! We need to modify the ast finder to search for template literals that are surrounded by gql, I can help with that!

@peterp I moved the Cell input typing stuff here, and setup the ast finder to search for gql queries. Do you want to take a look if this seems like the right direction to you?

**Comment by peterp** [comment]
Oooh, this is looking great, I'll review asap!

File: packages/internal/src/generate/typeDefinitions.ts [hunk 2]
```diff
@@ -103,7 +106,18 @@ export const generateMirrorCell = (p: string, rwjsPaths = getPaths()) => {
 
   const typeDefPath = path.join(mirrorDir, typeDef)
   const { name } = path.parse(p)
-  writeTemplate('templates/mirror-cell.d.ts.template', typeDefPath, { name })
+
+  const fileContents = fs.readFileSync(p, 'utf-8')
+  const gqlQueries = getGqlQueries(fileContents)
+
+  // For mirror cells we only care about the first gqlQuery
+  const gqlDoc = parseGqlQueryToAst(gqlQueries[0])[0]
```
Comment (peterp) [inline packages/internal/src/generate/typeDefinitions.ts:114]: @dac09 @jtoar I think I see a bit of a gap in the api-design of these functions here.

It's "sort of true" that we care about the first GQL query, but what we really care about is the GraphQL that has a named export with the identifier of "QUERY".

I'm happy for this to go in, but I think I would like to take a stab at making the api a bit more robust afterwards.

File: packages/internal/src/__tests__/gql.test.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import gql from 'graphql-tag'
```
Comment (peterp) [inline packages/internal/src/__tests__/gql.test.ts:1]: Can we add this as a dependency of internal?

**Review by peterp — APPROVED** [review]

File: packages/internal/src/__tests__/gql.test.ts (hunk 3 above)
Comment (peterp) [inline packages/internal/src/__tests__/gql.test.ts:1]: Can we add this as a dev-dependency?

File: packages/internal/src/generate/typeDefinitions.ts (hunk 2 above)
Comment (dac09) [reply packages/internal/src/generate/typeDefinitions.ts:114]: You're right, I'll see what I can do!

Comment (dac09) [reply packages/internal/src/generate/typeDefinitions.ts]: @peterp updated PR now... took me a while to find a good way to do this, but I think this implementation is a lot more solid. Babel with types is really hard man!

File: packages/internal/src/__tests__/gql.test.ts (hunk 3 above)
Comment (jtoar) [reply packages/internal/src/__tests__/gql.test.ts:1]: Added here: https://github.com/redwoodjs/redwood/pull/2756/files#diff-5babe0fba96044e558791b7fe897b7db1652789839de50f93e34cbe41efc636fR37

File: packages/internal/src/ast.ts [hunk 4]
```diff
@@ -59,6 +59,56 @@ export const getNamedExports = (code: string): NamedExports[] => {
...
 
+/**
+ * get all the gql queries from the supplied code
+ */
+export const getGqlQueries = (code: string) => {
+  const gqlQueries: string[] = []
+  const ast = parse(code) as types.Node
+  traverse(ast, {
+    TaggedTemplateExpression(path) {
+      console.log(path)
```
Comment (peterp) [inline packages/internal/src/ast.ts:70]:
```suggestion
```

**Review by peterp — COMMENTED** [review]
So cool! Thanks for this gents!

File: packages/internal/src/generate/typeDefinitions.ts (hunk 2 above)
Comment (peterp) [reply packages/internal/src/generate/typeDefinitions.ts]: The wallaby really helps here!

Comment (peterp) [reply packages/internal/src/generate/typeDefinitions.ts]: This is awesome btw, thanks for doing it!

Comment (dac09) [reply packages/internal/src/generate/typeDefinitions.ts]: Thank you for showing me how to use it, I introduced it to Dom the other day too. I think I just needed that little nudge to start using Wallaby properly, been and absolutely godsend for stuff like this

**Pull request merged**